pulp_docker/app/token_verification.py

```diff
@@ -0,0 +1,155 @@
+        if not self.is_token_valid(token):
+            authenticate_header = self._build_authenticate_header()
+            raise web.HTTPUnauthorized(
+                headers={'Www-Authenticate': authenticate_header},
+                reason='Access to the requested resource is not authorized. '
+                       'A provided Bearer token is invalid.'
+            )
+
+    def _get_token(self, authorization):
+        """
+        Get a raw token string from the header.
+
+        This method returns a string that skips the keyword 'Bearer' and an additional
+        space from the header (e.g. "Bearer abcdef123456" -> "abcdef123456").
+        """
+        return authorization[7:]
+
+    def _build_authenticate_header(self):
+        """
+        Build a header that a registry returns as a response to unauthorized access.
+
+        The method creates a value for a Www-Authenticate header. This value is used
+        by a client for requesting a Bearer token from a token server.
+        """
+        source_path = self._get_current_content_path()
+        return self._build_authenticate_string(source_path)
+
+    def _build_authenticate_string(self, source_path):
+        """
+        Build a formatted authenticate string.
+
+        For example, A created string is the following format:
+        realm="https://token",service="docker.io",scope="repository:my-app:push".
+        """
+        scope = f'repository:{source_path}:pull'
+        realm = f'{self.request.scheme}://{settings.TOKEN_SERVER}'
+        return f'Bearer realm="{realm}",service="{settings.CONTENT_HOST}",scope="{scope}"'
+
+    def is_token_valid(self, encoded_token):
+        """Decode and validate a token."""
+        with open(settings.PUBLIC_KEY_PATH, 'rb') as public_key:
+            decoded_token = self.decode_token(encoded_token, public_key.read())
+
+        return self.contains_accessible_actions(decoded_token)
+
+    def decode_token(self, encoded_token, public_key):
+        """
+        Decode token and verify a signature with a public key.
+
+        If the token could not be decoded with a success, a client does not have a
+        permission to operate with a registry.
+        """
+        jwt_config = self._init_jwt_decoder_config()
+        try:
+            decoded_token = jwt.decode(encoded_token, public_key, **jwt_config)
+        except jwt.exceptions.InvalidTokenError:
+            decoded_token = {'access': []}
+        return decoded_token
+
+    def _init_jwt_decoder_config(self):
+        """Initialize a basic configuration used for sanitizing and decoding a token."""
+        return {
+            'algorithms': [settings.TOKEN_SIGNATURE_ALGORITHM],
+            'issuer': settings.TOKEN_SERVER,
+            'audience': settings.CONTENT_HOST
+        }
+
+    def contains_accessible_actions(self, decoded_token):
+        """Check if a client has an access permission to execute the pull/push operation."""
```

Comment:
what about checking o token expiration? is it handled inside of jwt.decode?

Reply:
Yes, it is handled by the library.